File: mobile/test/money_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:offline_wallet/core/money.dart';
import 'package:offline_wallet/core/result.dart';

void main() {
  group('Money (ADR-4: integer paise, INR)', () {
    test('constructs from valid integer paise', () {
      final r = Money.fromPaise(500);
      expect(r.isOk, isTrue);
      final m = (r as Ok).value;
      expect(m.paise, 500);
      expect(m.currency, 'INR');
    });

    test('rejects negative amounts', () {
      expect(Money.fromPaise(-1).isErr, isTrue);
    });

    test('converts whole rupees to paise', () {
      final r = Money.fromRupees(5) as Ok;
      expect(r.value.paise, 5 * kPaisePerRupee);
    });

    test('adds without rounding error', () {
      final a = (Money.fromPaise(1) as Ok).value;
      final b = (Money.fromPaise(2) as Ok).value;
      expect(a.add(b).paise, 3);
    });

    test('subtract fails rather than going negative', () {
      final a = (Money.fromPaise(100) as Ok).value;
      final b = (Money.fromPaise(200) as Ok).value;
      expect(a.subtract(b).isErr, isTrue);
    });

    test('formats as INR rupees', () {
      expect((Money.fromPaise(500) as Ok).value.format(), '₹5.00');
    });
  });
}

File: backend/tests/money.test.ts
import { Money, PAISE_PER_RUPEE } from '../src/shared/money';
import { isOk, isErr, unwrap } from '../src/shared/result';

describe('Money (ADR-4: integer paise, INR)', () => {
  it('constructs from valid integer paise', () => {
    const r = Money.fromPaise(500);
    expect(isOk(r)).toBe(true);
    const m = unwrap(r);
    expect(m.paise).toBe(500);
    expect(m.currency).toBe('INR');
  });

  it('rejects non-integer paise (no floats in the ledger)', () => {
    expect(isErr(Money.fromPaise(1.5))).toBe(true);
  });

  it('rejects negative amounts', () => {
    expect(isErr(Money.fromPaise(-1))).toBe(true);
  });

  it('converts whole rupees to paise', () => {
    expect(unwrap(Money.fromRupees(5)).paise).toBe(5 * PAISE_PER_RUPEE);
  });

  it('adds without rounding error', () => {
    const a = unwrap(Money.fromPaise(1));
    const b = unwrap(Money.fromPaise(2));
    expect(a.add(b).paise).toBe(3);
  });

  it('subtract fails rather than going negative', () => {
    const a = unwrap(Money.fromPaise(100));
    const b = unwrap(Money.fromPaise(200));
    expect(isErr(a.subtract(b))).toBe(true);
  });

  it('formats as INR rupees', () => {
    expect(unwrap(Money.fromPaise(500)).format()).toBe('₹5.00');
  });
});
